pkg/queue/sharedmain/main.go

```diff
@@ -110,23 +104,77 @@ type config struct {
 	// Concurrency State Endpoint configuration
 	ConcurrencyStateEndpoint  string `split_words:"true"` // optional
 	ConcurrencyStateTokenPath string `split_words:"true"` // optional
+
+	Env
+}
+
+// Env exposes parsed QP environment variables for use by Options (QP Extensions)
+type Env struct {
+	// ServingNamespace is the namespace in which the service is defined
+	ServingNamespace string `split_words:"true" required:"true"`
+
+	// ServingService is the name of the service served by this pod
+	ServingService string `split_words:"true"` // optional
+
+	// ServingConfiguration is the name of service configuration served by this pod
+	ServingConfiguration string `split_words:"true" required:"true"`
+
+	// ServingRevision is the name of service revision served by this pod
+	ServingRevision string `split_words:"true" required:"true"`
+
+	// ServingPod is the pod name
+	ServingPod string `split_words:"true" required:"true"`
+
+	// ServingPodIP is the pod ip address
+	ServingPodIP string `split_words:"true" required:"true"`
 }
 
+// Defaults provides Options (QP Extensions) with the default bahaviour of QP
+// Some attributes of Defaults may be modified by Options
+// Modifying Defaults mutates the behavior of QP
+type Defaults struct {
+	// Logger enables Options to use the QP pre-configured logger
+	// It is expected that Options will use the provided Logger when logging
+	// Options should not modify the provided Default Logger
+	Logger *zap.SugaredLogger
+
+	// Env exposes parsed QP environment variables for use by Options
+	// Options should not modify the provided environment parameters
+	Env Env
+
+	// Ctx provides Options with the QP context
+	// An Option may derive a new context from Ctx. If a new context is derived,
+	// the derived context should replace the value of Ctx.
+	// The new Ctx will then be used by other Options (called next) and by QP.
+	Ctx context.Context
+
+	// Transport provides Options with the QP RoundTripper
+	// An Option may wrap the provided Transport to add a Roundtripper.
+	// If Transport is wrapped, the new RoundTripper should replace the value of Transport.
+	// The new Transport will then be used by other Options (called next) and by QP.
+	Transport http.RoundTripper
+}
+
+type Option func(*Defaults)
+
 func init() {
 	maxprocs.Set()
 }
 
-func Main() {
-	ctx := signals.NewContext()
+func Main(opts ...Option) error {
+	d := Defaults{
+		Ctx: signals.NewContext(),
+	}
 
 	// Parse the environment.
 	var env config
 	if err := envconfig.Process("", &env); err != nil {
-		fmt.Fprintln(os.Stderr, err)
```

Comment:
We no longer print to `stderr` after failing to read the env variables

Reply:
We should add this back

Reply:
Done